querqy-core/src/main/java/querqy/rewrite/ContextAwareQueryRewriter.java

```diff
@@ -1,39 +1,118 @@
 /**
- * 
+ *
  */
 package querqy.rewrite;
 
+import querqy.infologging.InfoLoggingContext;
+import querqy.model.AbstractNodeVisitor;
 import querqy.model.ExpandedQuery;
+import querqy.model.Node;
+
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
 
 /**
  * If a {@link QueryRewriter} implements this interface, the {@link #rewrite(ExpandedQuery, SearchEngineRequestAdapter)}
  * method is called instead of {@link #rewrite(ExpandedQuery)} in the query rewriting chain. The Map argument can be
- * used to pass data between rewriters and to the consumer of the rewrite chain.
- * 
+ * used to pass data between rewriters and to the consumer of the rewrite chain. It also gives you the easy possibility to
+ * track the applied rules in a rewriter.
+ *
  * @author René Kriegler, @renekrie
  *
  */
-public interface ContextAwareQueryRewriter extends QueryRewriter {
```

Comment:
Unfortunately this is too much of a change: turning the interface into an abstract class means that all custom rewriter implementations out there need to be changed.

How about this: leave ContextAwareQueryRewriter as an interface and create a new abstract class (AbstractLoggingRewriter??) that implements the interface and that looks more or less like your abstract ContextAwareQueryRewriter. You'd let the rewriters that come with Querqy implement it, and only those custom rewriters that use info logging need to be changed (which is probably very few)

Reply:
But see my note below re making the SearchEngineRequestAdapter a member variable.

Reply:
We (Tobi and I)  restored the interface and extracted the code to an abstract class like you suggested.

The SearchEngineRequestAdapter already was a member variable of the CommonRulesRewriter before so I moved it from the abstract class to both the CommonRules and the ReplaceRewriter.

What do you think about these changes?